pkgs/by-name/as/astro-language-server/package.nix

```diff
@@ -0,0 +1,97 @@
+{ lib
+, stdenv
+, fetchFromGitHub
+, stdenvNoCC
+, jq
+, moreutils
+, nodePackages
+, cacert
+}:
+
+stdenv.mkDerivation rec {
+  pname = "astro-language-server";
+  version = "2.8.4";
+
+  src = fetchFromGitHub {
+    owner = "withastro";
+    repo = "language-tools";
+    rev = "@astrojs/language-server@${version}";
+    hash = "sha256-pUo8bZPQSYLKE9aZHECXOCxlUjiGRdECkqPClRVVick=";
+  };
+
+  sourceRoot = "${src.name}/packages/language-server";
+
+  pnpm-deps =
+    stdenvNoCC.mkDerivation {
+      pname = "${pname}-pnpm-deps";
+      inherit src version;
+
+      nativeBuildInputs = [
+        jq
+        moreutils
+        nodePackages.pnpm
+        cacert
+      ];
+
+      # https://github.com/NixOS/nixpkgs/blob/de80f1eeac3152c5bbfb1f8891b6414d526bfc54/pkgs/by-name/po/pot/package.nix#L54
+    installPhase = ''
+      export HOME=$(mktemp -d)
+      pnpm config set store-dir $out
+      # use --ignore-script and --no-optional to avoid downloading binaries
+      # use --frozen-lockfile to avoid checking git deps
+      pnpm install --frozen-lockfile --no-optional --ignore-script
+
+      # Remove timestamp and sort the json files
+      rm -rf $out/v3/tmp
+      for f in $(find $out -name "*.json"); do
+        sed -i -E -e 's/"checkedAt":[0-9]+,//g' $f
+        jq --sort-keys . $f | sponge $f
+      done
+    '';
+
+      dontFixup = true;
+      outputHashMode = "recursive";
+      outputHash = "sha256-JkLgntrZLUs4JgrHxJLyWfkLUEuORyYJqJAfoA1M2Cs=";
+    };
+  nativeBuildInputs = [
+    nodePackages.pnpm
+    nodePackages.nodejs
+  ];
+
+  preBuild = ''
+    export HOME=$(mktemp -d)
+    pnpm config set store-dir ${pnpm-deps}
+    pnpm config set recursive-install false
+
+    chmod -R +w ../..
+
+    pnpm install --offline --frozen-lockfile --no-optional --ignore-script
+    patchShebangs node_modules/{*,.*}
+  '';
+
+  buildPhase = ''
+    runHook preBuild
+
+    pnpm build
+
+    runHook postBuild
+  '';
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir $out
+    pnpm deploy --filter=language-server --offline --prod --no-optional $out
+
+    mv $out/bin/nodeServer.js $out/bin/astro-ls
+
+    runHook postInstall
+  '';
+
+  meta = {
+    description = "The Astro language server";
+    homepage = "https://github.com/withastro/language-tools";
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [pyrox0];
+  };
```

Comment:
Please set `meta.mainProgram` to `"astro-ls"`.

Reply:
Both of these done